src/backend/compatIO.test.js

```diff
@@ -0,0 +1,212 @@
+// @flow
+
+import tmp from "tmp";
+import path from "path";
+import fs from "fs-extra";
+import {type Compatible, toCompat, fromCompat} from "../util/compat";
+import {createProject, projectToJSON, projectFromJSON} from "../core/project";
+import {Graph} from "../core/graph";
+import * as WeightedGraph from "../core/weightedGraph";
+import {node as graphNode} from "../core/graphTestUtil";
+import {compatWriter, compatReader} from "./compatIO";
+
+type MockDataType = {+[string]: string};
+
+const mockCompatInfo = {type: "mock", version: "1"};
+
+function mockToJSON(x: MockDataType): Compatible<MockDataType> {
+  return toCompat(mockCompatInfo, x);
+}
+
+function mockFromJSON(x: Compatible<any>): MockDataType {
+  return fromCompat(mockCompatInfo, x);
+}
+
+describe("src/backend/compatIO", () => {
+  describe("compatWriter", () => {
+    it("should work with example type", async () => {
+      // Given
+      const filePath = path.join(tmp.dirSync().name, "mockData.json");
+      const data: MockDataType = {foo: "bar"};
+
+      // When
+      const writer = compatWriter(mockToJSON);
+      await writer(filePath, data);
+
+      // Then
+      const actualData = await fs.readFile(filePath, "utf8");
+      expect(actualData).toEqual(
+        `[{"type":"mock","version":"1"},{"foo":"bar"}]`
+      );
+    });
+
+    it("should handle write errors", async () => {
+      // Given
+      const filePath = tmp.dirSync().name;
+      const data: MockDataType = {foo: "bar"};
+
+      // When
+      const writer = compatWriter(mockToJSON);
+      const p = writer(filePath, data);
+
+      // Then
+      await expect(p).rejects.toThrow("Could not write data:\nError: EISDIR");
+    });
+
+    it("should accept a typeName to improve errors", async () => {
+      // Given
+      const filePath = tmp.dirSync().name;
+      const data: MockDataType = {foo: "bar"};
+      const name = "TestName";
+
+      // When
+      const writer = compatWriter(mockToJSON, name);
+      const p = writer(filePath, data);
+
+      // Then
+      await expect(p).rejects.toThrow(
+        `Could not write ${name} data:\nError: EISDIR`
+      );
+    });
+
+    it("should use stable json serialization", async () => {
+      // Given
+      const filePath = path.join(tmp.dirSync().name, "mockData.json");
+      const data: MockDataType = {bbb: "second", aaa: "first"};
+
+      // When
+      const writer = compatWriter(mockToJSON);
+      await writer(filePath, data);
+
+      // Then
+      const actualData = (await fs.readFile(filePath)).toString("utf-8");
+      expect(actualData).toEqual(
+        `[{"type":"mock","version":"1"},{"aaa":"first","bbb":"second"}]`
+      );
+    });
+  });
+
+  describe("compatReader", () => {
```

Comment:
I'd like to see a test that shows that we get a helpful error message on compat version mismatch.

Reply:
Added one for type and version mismatch.